Add low level intro to CI (#4)
lyashevska/ds-cr-slides: ci/00_title.md

</style>

# Continuous Integration (CI)

<div class="blockquote-wrapper fragment">
  <div class="blockquote">

>      It is the practice of integrating changes to a project made by individuals into a main, shared version frequently.

Reviewer: How about this?

```suggestion
     means keeping the main (shared) version of a project up to date via frequent or automated integration of changes.
```

Author: I like it more! Although we cannot quote the turing way on this. Can I quote you on this? :D

    <h4>&mdash;https://book.the-turing-way.org</h4>
  </div>
</div>

---

<!-- .slide: data-state="blue_overlay yellow_flag yellow_strip purple_half_circle_bottom purple_blob right_e_top" data-background-video="./files/steampunk-85358.mp4" data-background-video-loop data-background-video-muted="true" data-auto-animate data-auto-animate-id="title" -->

# Motivation

> How can we implement __automatic testing__ each time we push changes to the repository? 

Reviewer: This sentence does not really fit the header "motivation".
Also, I find that in the font we are using, the boldface is so similar to standard text, that rather than creating emphasis, it becomes distracting, so I don't like using it.
Maybe I'll open a separate issue about this.

Author: ```suggestion
# Motivation
We want to automatically test our code each time we push changes to the repository.
```

Author: How about it?

<img src="./files/automate.jpg">

---

<!-- .slide: data-state="blue_overlay yellow_flag logo 9" data-background="./files/reload-97640.svg" data-background-size="50%" -->

> ## CI enables automatic testing!

Reviewer: I don't think this is accurate, at least not according to definition above. I think "Continuous Testing" is a thing in itself, which is separate from (but related to) CI.

GitHub/GitLab actions or workflows enable automated testing.

Also, I kind of like the explanations/definitions given on [wiki](https://en.wikipedia.org/wiki/CI/CD)

Author: Hmm maybe, in my head Automated testing is a subset of CI. No?

Author: Oh I see that according to wiki continuous integration and continuous testing are separate things.. So integration is only integration and testing comes after.. Very confusing. We are teaching CI but need to talk about testing (their exercise is about testing). How would you rephrase it then?

Reviewer: So, I think to me the main problem is that the way you talk about CI throughout the module does not fit the definition we give at the beginning.

Maybe one way around this is to introduce it in the first slide as: "Formally speaking, CI is defined as [XXX]. However, in daily practice, we often use this terminology in a broader sense, namely as [YYY]." 
I do think in this case we need to think about a good "definition" or description of [YYY], so that it's clear from the beginning what we're all talking about and everyone is on the same page.

Allows to identify any conflicts and bugs that are introduced by changes, so they are found and fixed early, minimising the effort required to do so. 

---

<!-- .slide: data-state="blue_overlay yellow_flag logo 9" data-background="./files/reload-97640.svg" data-background-size="50%" -->

## What else can CI do?
- Building & compiling

> - Documentation, ...

Reviewer: I think this made sense to be indented, it was examples of things that can be auto-built/compiled.

You can't have workflows write documentation for you (unfortunately), which is what it looks like now.

Author: Yes of course, nice.

- Deploying (PyPi, Kubernetes, GitHub Pages)
  - Just like these slides

> - Linting (analyze code for potential errors)

Reviewer: ```suggestion
- Code analysis
    - linting (style conventions), formatting, quality metrics, ...
```

- Security analyses
- Send messages
  - Slack, Discord, Matrix, Mastodon, email, ...

---

<!-- .slide: data-state="blue_overlay yellow_flag logo 9" data-background="./files/reload-97640.svg" data-background-size="50%" -->

> ## CI service providors

Reviewer: Maybe also mention that there are defaults that people can download and adapt for this and that many near plug-and-play examples can also be found online.

Author: Yes, when I present it I will tell.
